File: warehouse/assembly.py
# coding: utf-8

import cStringIO, os.path, logging, Image, ImageFilter
from core.sw_base import BaseHandler
from avtuk.avtuk_models import Header, PastingType, StorageType
import settings as config
from barcode.zek_model import mm2pix

import urlparse

from avtuk.executor import Executor

from tornado.httpclient import HTTPError


from tornado import template
import cx_Oracle

from settings import ROOT_DIR


from core.common_tools import fetchone, fetchall

#=== Сборка ==================================================================#
class AssemblyHandler(BaseHandler):
    urls = r'/warehouse/assembly'
    
 

    def get(self):
 

        opers = [{"id":"opr%s" % k, "type":"button", "text":"%s" % v, "action":"do_tool_3"}
                 for k, v in enumerate(('-',
                                        'Без розничных продаж и подарков',
                                        'Только розничные продажи',
                                        'Только подарки клиентам'))]
        opers[0]['selected'] = "true"
        
        pasttypes = PastingType.select();
        
        storagetypes = StorageType.select();
        
        
        
        pDepart_type = 3 #pDepart_type – тип департаментов,  в данном случае константа 3 (сборщики)
        
        out = self.cursor.var(cx_Oracle.CURSOR)    
        res = self.proc("shiva.GetSotrudList", [self.session.rc, pDepart_type, out])
                
        all_sotrud = res[-1].fetchall()
   
        
        prn = [{"id":"prn%s" % k, "type":"button", "text":"%s" % v, "action":"do_tool_4"}
                 for k, v in enumerate(('Накладная',
                                        'Паспорта качества',
                                        'Отгрузочные этикетки',
                                        'Упаковочный лист',
                                        'Сборочные'))]


        # renderSelect - do not remember last operation in buttonSelect list
 
        loader = template.Loader(os.path.join(ROOT_DIR, 'warehouse'))
        result = loader.load("assembly.js").generate(pasttypes=pasttypes, storagetypes=storagetypes, all_sotrud=all_sotrud)
       
 
 
        self.write({'def':[ 
                              
                           {'id':"btnactions", 'type':"buttonSelect", 'text':'Действия', 'openAll': 'true', 'renderSelect': 'false', 'items':
                            [                             
                             {"id":"act1", 'type':"button", 'text':'Подтвердить', 'img':"accept24.png", 'imgdis':"accept24g.png", 'action':'do_tool_save'},
                             {"id":"act2", "type":"button", "text":"Консолидировать", "action":"do_tool_consolidation"},
                             
                             {'type':"separator"},
                             {"id":"act4", 'type':"button", 'text':'Доп.информация', 'title':'Ввод дополнительной информации', 'img':"add24.png", 'action':'do_tool_edit_form'},
                             {"id":"act5", 'type':"button", 'text':'Печать паспорта качества', 'title':'Печать паспорта качества', 'img':"print.gif", 'imgdis':"print_dis.gif", 'action':'do_tool_4'},
                             
                             {'type':"separator"},
                             {"id":"act3", "type":"button", "text":"Расконсолидировать", "action":"do_tool_deconsolidation"},
                             {'type':"separator"},
                             {"id":"act7", "type":"button", "text":"Откатить фактуру", "action":"do_tool_rollback"}
                            ]},
                           
                           {'type':"separator"},
                           {'type':"text", 'text':"  Фильтр:"},
                           {'id':"btnopr", 'type':"buttonSelect", 'openAll': 'true', 'renderSelect': 'false', 'text':'-', 'items':opers},
                                              
                           {'type':"separator"},
                           {'type':"separator"},
                           {'id':"btnprn", 'type':"buttonSelect", 'text':'Печать', 'items':prn},

                           {'type':"separator"},
                           {'type':"button", 'img':"excel24.png", 'imgdis':"excel24g.png", 'action':'do_tool_csv', 'title':'Сохранить в CSV'},
                           
                           {'type':"separator"},
                           {'type':"button", 'action':'do_tool_test', 'text':'TEST', 'title':'TEST'}],



                    'one':[{'type':"button", 'text':'Сохранить', 'img':"accept24.png", 'imgdis':"accept24g.png", 'action':'do_save_extinfo'},
                           {'type':"button", 'text':'Отмена', 'img':"delete24.png", 'imgdis':"delete24g.png", 'action':'do_cancel_extinfo'},
                           '<div id="swformextinfo" />'],
                    'two':[{'type':"button", 'text':'Назад', 'img':"delete24.png", 'imgdis':"delete24g.png", 'action':'do_cancel_extinfo'},
                           ],
                    'cmd': result
                    })



#==============================================================================
class AssemblyDataHandler(BaseHandler):
    urls = r'/warehouse/assembly/data/([^/]+)'

    def get(self, param):
        w, h = 200, 142
        
                 
        head  = self.get_argument("head", 0)
        fname = self.get_argument("fname", '')
       
        header_id = self.get_argument("id", default=None)
        
        
        if param == 'item':

            out = self.cursor.var(cx_Oracle.CURSOR)    
            res = self.proc("shiva.GetHeaderAssemblyList", [header_id, None, None, None, out])
                
            all_results = fetchone(res[-1])        
                
            self.write(all_results)
            return        
        
               

        if param == 'head':

            pFilter = self.get_argument("oper", None)
            
            sotrud = self.session.uid
            
            out = self.cursor.var(cx_Oracle.CURSOR)    
            
            res = self.proc("shiva.GetHeaderAssemblyList", [None, self.session.rc, pFilter, sotrud, out])
                
            all_results = fetchall(res[-1], count=0)
                
            self.write(all_results)
            return
                

        elif param == 'info':
         
            self.write(Header.get_item(head, rc=self.session.rc).as_json('prim', 'marsh_cls.name')[0])
            # self.write({'error': 'test' })
        
            
            
        elif param == 'messages':

            sql = "select t.code,t.name,f.valume from factura f left join tovar t on isclad.GetTovarFromModify(f.tovar)=t.id where f.header=:pHeader"

            self.write_XML(Executor.exec_cls(sql, pHeader=head)
                           .as_grid('code', 'name', 'valume'))
                        

        elif param == 'tovar':

            sql = "select t.code,t.name,f.valume from factura f left join tovar t on isclad.GetTovarFromModify(f.tovar)=t.id where f.header=:pHeader"

            self.write_XML(Executor.exec_cls(sql, pHeader=head)
                           .as_grid('code', 'name', 'valume'))

    def post(self, param):
        
        header_id = self.get_argument("ids", default=None)
        pOkNoParty = self.get_argument("pOkNoParty", default=0)
            
        if param == 'gts':                                 
            
            self.proc("shiva.AddShedulerHeader", [header_id, pOkNoParty, self.session.uid])            
            #self.write({'info':'Сформированы задания на сборку'})
            
            return
        


        elif param == 'save':

            header_id = self.get_argument("id",    default=None)
            skin    = self.get_argument("skin",    default=None)
            desk    = self.get_argument("desk",    default=None)
            prim    = self.get_argument("prim",    default=None)
            pallet  = self.get_argument("pallet",  default=None)
            os_numb = self.get_argument("os_numb", default=None)            
            storage = self.get_argument("storage", default=None)
            
            packing = self.get_argument("packing", default=0)
            sotrud1 = self.get_argument("sotrud1", default=None)
            sotrud2 = self.get_argument("sotrud2", default=None)
            
            
            out = self.cursor.var(cx_Oracle.STRING)    
            res = self.proc("shiva.SetHeaderInfo", [header_id,
                                                               pallet,
                                                               os_numb,
                                                               prim,
                                                               skin,
                                                               storage,
                                                               desk,
                                                               packing,
                                                               sotrud1,
                                                               sotrud2,
                                                               out])
                
            result = str(res)

            self.write({'info':'Информация сохранена'})
   





        elif param == 'consolidation':            


            out = self.cursor.var(cx_Oracle.STRING)    
            res = self.proc("tsclad.MakeTotalHeaderNew", [header_id, out])
                
            result = res[-1]
            
            self.write({'info':'Консолидация выполнена: фактуры: %s' % result, 'ids': result})



        elif param == 'deconsolidation':            

            result = self.proc('tsclad.DelTotalHeaderSclad', [header_id])       
            self.write({'info':'Расконсолидация выполнена успешно'})

                
        elif param == 'rollback':            
            
            result = self.proc('shiva.RollbackHeader', [header_id, cx_Oracle.NUMBER])       

            self.write({'info':'Откат фактуры выполнен успешно'})
               

File: warehouse/assembly.js
var sw_btk = new dhtmlXLayoutObject(self.Panels["def"], "2E");

//sw_btk.setAutoSize("b;c", "a;c");

sw_btk.items[0].setText("Список фактур");
sw_btk.items[1].setText("Список товаров");
window.Cleaner.push(sw_btk);
                            
var sw_grid1 = sw_btk.cells("a").attachGrid(1000);

sw_grid1.enableAutoWidth(true);
sw_grid1.setStyle("width: 100% !important;", "");

sw_grid1.columns([
	  { type:"ro", sort:"int",  align:"right", width:"66", id:"5555555",  label:["ID", "#text_filter"] },
	  { type:"ro", sort:"int",  align:"right", width:"45",  label:["Номер", "#text_filter"] },
	  { type:"ro", sort:"date", align:"right", width:"66",  label:["Дата", "#text_filter"] },
	  { type:"ro", sort:"str",  align:"left",  width:"276", label:"Получатель"},
	  { type:"ro", sort:"str",  align:"left",  width:"286", label:"Операция" },
	  { type:"ro", sort:"str",  align:"left",  width:"76",  label:["Статус", "#select_filter"]},
	  { type:"ro", sort:"str",  align:"left",  width:"76",  label:["N отправки"]},
	  { type:"ro", sort:"str",  align:"left",  width:"0"},
	  
	  { type:"ro", sort:"str",  align:"left",  width:"96",  label:["Оклейка", "#select_filter"]},
	  
	  { type:"ro", sort:"str",  align:"left",  width:"10"},
	  { type:"ro", sort:"str",  align:"left",  width:"10"}
]);
                            
sw_grid1.enableMultiselect(true);

                                
window.gts = 1; // ворота
window.ids = 0;


window.Cleaner.push(sw_grid1);
window.do_tool_csv = function(){ self.GridCSV(sw_grid1) }





var pasttypes = [{value: "", text: ""},
	              {% for item in pasttypes %}
					 {value: "{{ item['id'] }}", text: "{{ item['name'] }}"},
				  {% end %}];

var storagetypes = [{value: "", text: ""},
             {% for item in storagetypes %}
				 {value: "{{ item['id'] }}", text: "{{ item['name'] }}"},
			  {% end %}];


var all_sotrud = [{value: "", text: ""},
                    {% for item in all_sotrud %}
       				 {value: "{{ item[0] }}", text: "{{ item[1]}}"},
       			  {% end %}];


var swForm = new dhtmlXForm("swformextinfo", [
	{type:"input",  name:"id",      bind:"id",      label:"id:", 	           labelWidth: 110, disabled:"true" },
    {type:"input",  name:"os_numb", bind:"os_numb", label:"Номер отправки:",   labelWidth: 110, value: ""},
    {type:"select", name:"pallet",  bind:"pallet",  label:"Тип оклейки:", 	   labelWidth: 110, options: pasttypes, disabled:"true" },
    {type:"select", name:"storage", bind:"storage", label:"Тип размещения:",   labelWidth: 110, options: storagetypes},
    {type:"input",  name:"prim",  	bind:"prim", 	label:"Примечание:",	   labelWidth: 110, value: ""},
    {type:"input",  name:"skin",    bind:"skin", 	label:"Кол-во пленки:",    labelWidth: 110, value: "0"},
    {type:"input",  name:"desk",  	bind:"desk", 	label:"№ первой коробки:", labelWidth: 110, value: ""},
    
    {type:"select", name:"sotrud1", bind:"sotrud1", label:"Сборщик 1:",      labelWidth: 110, options: all_sotrud},
    {type:"select", name:"sotrud2", bind:"sotrud2", label:"Сборщик 2:",      labelWidth: 110, options: all_sotrud},
   
    {type:"checkbox", name:"packing", bind:"packing", label: "Пленка на коробки",            labelWidth: 110, checked: false}
]);


var sw_os_numb = swForm.getInput("os_numb");
var sw_prim = swForm.getInput("prim");

window.Cleaner.push(swForm);
            





var sw_grid2 = sw_btk.cells("b").attachGrid();
sw_grid2.columns([
  
      { type:"ro", sort:"int",  align:"right", width:"55", label:"Код" },
      { type:"ro", sort:"str",  align:"left",  width:"*",  label:"Товар" },
      { type:"ro", sort:"int",  align:"right", width:"65", label:"Кол-во" }
      
]);
window.Cleaner.push(sw_grid2);


sw_grid1.attachEvent("onRowSelect", function(id){

        self._messager.clearAll();


        self.LoadGrid(sw_grid2, "/warehouse/assembly/data/tovar?head="+id);

        self.NetSendAsync("/warehouse/assembly/data/info?head="+id, false,function(resp) {                            
                                                   
            if (resp.marsh_cls_name){                              
                self.AddMessage(resp.marsh_cls_name,5);
            }
            if (resp.prim){                              
                self.AddMessage(resp.prim,5);
            }
                                            
            return;
        
        });
        
        
        self.NetSendAsync("/warehouse/sborka/data/messages?head="+id, false,function(resp) {                            
            
        	for (var i=0; i<resp.length; i++) {
        	    // Iterates over numeric indexes from 0 to 5, as everyone expects.
        		
        		self.AddMessage(resp[i].mess,5);
        	}
                                            
            return;
        
        });
                         
});




window.do_tool_test = function(){


	
};



window.do_tool_edit_form = function(ids){
    
	
	ids=sw_grid1.getSelectedRowId();;
    
    swForm.clear();
    swForm.lock();
    
    
    dhtmlxAjax.get("/warehouse/assembly/data/item?" + encodeURI("id="+ ids), function(loader) {
    	
    	
    	var all_results =JSON.parse(loader.xmlDoc.responseText);
    	
		for (var item in all_results) {

    		try{
    			
    			swForm.setItemValue(item, all_results[item]);
    			
    			// console.log( all_results[item] );
    			
    		}
    		catch(e){
    			
    		}
    	}
    	
    
    	
    	swForm.unlock();
    	window.swForm = swForm;

	});
    	
                                
    
    self.ShowPanel("one");

    
                            
   
}



//запись доп. информации
window.do_save_extinfo = function(){

/*
 if (sw_os_numb.value.length < 1 || sw_os_numb.value.length > 32) {
     app.AddMessage('Недопустимая длина номера отправки',2);
     sw_os_numb.focus();
     return 0;
 }
*/  
 // Костыль к DHTMLX, иначе сериализация вручную
 swForm.setItemValue("os_numb", sw_os_numb.value);
 swForm.setItemValue("prim", sw_prim.value);
 
 var dt = swForm.Serialize();
 if(!!swForm.ids){
     dt=dt+"&uid="+swForm.ids;
 }
 //console.log('Serialized: ', dt);
 
 
 self.NetSendAsync("/warehouse/assembly/data/save", dt,function(resp) {
       
	 self.ShowPanel("def");
     
	 
	 window.do_tool_3();
      
  });
     
 
 self.ShowPanel("def");
  
}


 


        // save
window.do_tool_save = function(ids2, check){


    var ids = sw_grid1.getSelectedRowId();    
    if (!ids && ids2){
        ids = ids2;
    }
    
                         
    var pOkNoParty = '&pOkNoParty=1';  //Игноррировать товары без срока годности
    
    /*         
    if (check){
        pOkNoParty = '&pOkNoParty=1';
    }
    */
    
    
    if (!ids){
        self.AddMessage('Фактура не выбрана',3);
        return;
    }
    sw_grid2.editStop();
    
    
    self.NetSendAsync("/warehouse/assembly/data/gts", "ids="+ids+"&gts="+window.gts+pOkNoParty,function(resp) {
    
    window.do_tool_3();                                                         
                             
   });       
                         
}

/************************ SELECT CREATED NEW ROW  ********************/

                     


window.do_clear_filters = function(grid){
    
    try
    {                                   
                                       
        for (var i=0; i < grid.filters.length; i++){                                    
            grid.filters[i][0].value = '';
            //console.log(grid.filters[i][0].value);                                
        }
        
        grid.callEvent("onFilterEnd");
    }
    catch(e){}                                                                   
}                            
                     

window.do_tool_select = function(sw_grid1,ids){

    sw_grid2.clearAll();


    
    self.load(sw_grid1, "/warehouse/assembly/data/head?oper="+window.ids,function() {

        sw_grid1.selectRowById(ids,true,true,true);
                                         
    });
    
};


window.do_tool_consolidation = function(){
    var ids = sw_grid1.getSelectedRowId();
    if (!ids){
        self.AddMessage('Фактуры не выбраны',3);
        return;
    }
    //console.log("selected ids: " + ids);
    
    window.do_clear_filters(sw_grid1);
    /*
    sw_grid2.editStop();
    self.NetSendAsync("/warehouse/assembly/data/consolidation", "ids="+ids,function(resp) {
        window.do_tool_select(sw_grid1,resp.ids);
    });

    */




    dhtmlxAjax.post("/warehouse/assembly/data/consolidation",  "ids="+ids, function(resp){

        var results = JSON.parse(resp.xmlDoc.responseText);


        window.do_tool_select(sw_grid1,results.ids || "");

        if (results.info){
            self.AddMessage(results.info, "info");

        }


    });





}


window.do_tool_deconsolidation = function(){
    var ids = sw_grid1.getSelectedRowId();
    if (!ids){
        self.AddMessage('Фактуры не выбраны',3);
        return;
    }
    //console.log("selected ids: " + ids);
    
    window.do_clear_filters(sw_grid1);

    /*
    sw_grid2.editStop();
    self.NetSendAsync("/warehouse/assembly/data/deconsolidation", "ids="+ids, func = function() {
        
        window.do_tool_3(); 
    });*/



    dhtmlxAjax.post("/warehouse/assembly/data/deconsolidation",  "ids="+ids, function(resp){

        var results = JSON.parse(resp.xmlDoc.responseText);

        window.do_tool_select(sw_grid1,results.ids || "");


        if (results.info){
            self.AddMessage(results.info, "info");

        }



    });


}                         


window.do_tool_rollback = function(){
    var ids = sw_grid1.getSelectedRowId();
    if (!ids){
        self.AddMessage('Фактуры не выбраны',3);
        return;
    }
    
    
    window.do_clear_filters(sw_grid1);
    
    sw_grid2.editStop();
    self.NetSendAsync("/warehouse/assembly/data/rollback", "ids="+ids, func = function() {
        
        
        window.do_tool_3(); 
        });
}           

                            

// ворота
window.do_tool_2 = function(ids){
    window.gts=parseInt(ids.substr(3));
    var text = self.Toolbars["def"].getListOptionText("btngts", ids);                                            
    self.Toolbars["def"].setItemText("btngts", text);
}


// print passport
window.do_tool_4 = function(ids2){
    var ids = sw_grid1.getSelectedRowId();
    if(!ids) self.AddMessage('Выберите фактуру',2)
    else {
        var m=ids2.substr(3);
        var d='';
        //Отгрузочные этикетки
        if(m==2){
            var d=prompt('Диапазон мест разделенные пробелом','').split(' ');
        }
        self.NetSend("/warehouse/printpassport/data/print?head="+ids+"&mode="+m+"&d="+d);
    }                                
}



// Дополнительная информация - переход в режим ввода
window.do_tool_5 = function(){
    var ids = sw_grid1.getSelectedRowId();
    swForm.ids = ids;
    
    //console.log("ids: " + ids);
    
    if(!ids) {
        self.AddMessage('Выберите фактуру',2);
        return;
    }
    // а тут надо совершить проверку поля status
    var s = Number(sw_grid1.cells(ids, 7).getValue());
    if (s >= 4) {
        self.AddMessage('Фактуру с этим статусом нельзя изменить', 2);
        return;
    }

    swForm.setItemValue("os_numb", sw_grid1.cells(ids, 6).getValue());

    var opt = sw_grid1.cells(ids, 9).getValue();
    //console.log(opt);
    if (!!opt) {
        swForm.setItemValue("past_type", opt);
    } else {
        swForm.setItemValue("past_type", 1);
    }
    
    
    var opt2 = sw_grid1.cells(ids, 10).getValue();
    
    //console.log(opt2);
    
    if (!!opt2) {
        swForm.setItemValue("storage_type", opt2);
    } else {
        swForm.setItemValue("storage_type", "");
    }                                
    
    
    self.ShowPanel("one");
}

// Выход из режима ввода доп.информации
window.do_cancel_extinfo = function(){self.ShowPanel("def");}



// фильтр
window.do_tool_3 = function(ids2){     

    if (ids2){
        sessionStorage.setItem("filter_opr_/warehouse/assembly/data/head", ids2);

      }
    else{  
        var opr = sessionStorage.getItem("filter_opr_/warehouse/assembly/data/head");
        ids2 = opr?opr:"opr0";
    }
      
    window.ids=ids2.substr(3);                                                        

    var text = self.Toolbars["def"].getListOptionText("btnopr", ids2);
    
                                                                                
    self.Toolbars["def"].setItemText("btnopr", text);
    

   
    
     sw_grid2.clearAll();
    
    
     self.load(sw_grid1, "/warehouse/assembly/data/head?oper="+window.ids, function(){
    	 
     });
     
     
    }
    
  
        
window.do_tool_3();

